File: src/javascript/algorithms/graph/shortestDistanceFromAllBuildings/shortestDistanceFromAllBuildings.js
// BFS Shortest path TC: O(k * m * n) SC: O(m * n)
function shortestDistance(grid) {
    const rows = grid.length, cols = grid[0].length;
    const distance = Array.from({ length: rows }, () => new Array(cols).fill(0));
    let walk = 0;
    let answer = Infinity;
    const directions = [[-1, 0], [0, 1], [1, 0], [0, -1]];

    for (let row = 0; row < rows; row++) {
        for (let col = 0; col < cols; col++) {
            if (grid[row][col] === 1) {
                const queue = [[row, col]];
                let head = 0;
                let level = 0;

                while (head < queue.length) {
                    const size = queue.length - head;

                    for (let i = 0; i < size; i++) {
                        const [r, c] = queue[head++];
                        for (const [dr, dc] of directions) {
                            const nr = r + dr, nc = c + dc;
                            if (nr < 0 || nr >= rows || nc < 0 || nc >= cols || grid[nr][nc] !== walk) {
                                continue;
                            }
                            grid[nr][nc]--;                              // mark visited
                            distance[nr][nc] += level + 1;               // accumulate distance
                            answer = Math.min(answer, distance[nr][nc]);  // update best answer
                            queue.push([nr, nc]);
                        }
                    }

                    level++;
                }

                walk--;
            }
        }
    }

    return answer === Infinity ? -1 : answer;
}

function deepCopy(grid) {
    return grid.map(row => [...row]);
}

function runTests() {
    const tests = [
        { grid: [[1,0,2,0,1],[0,0,0,0,0],[0,0,1,0,0]], expected: 7  },
        { grid: [[1,0]],                                expected: 1  },
        { grid: [[1]],                                  expected: -1 },
        { grid: [[0,1,0],[0,0,0],[0,1,0]],              expected: 2  },
        { grid: [[0,2,1]],                              expected: -1 },
    ];

    for (const { grid, expected } of tests) {
        const result = shortestDistance(deepCopy(grid));
        const status = result === expected ? "PASS" : "FAIL";
        console.log(`[${status}] result=${result} expected=${expected}`);
    }
}

runTests();

File: src/python/algorithms/graph/shortestDistanceFromAllBuildings/shortest_distance_from_all_buildings.py
from collections import deque


# BFS Shortest path TC: O(k * m * n) SC: O(m * n)
def shortest_distance(grid: list[list[int]]) -> int:
    rows, cols = len(grid), len(grid[0])
    distance = [[0] * cols for _ in range(rows)]
    walk = 0
    answer = float('inf')
    directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    for row in range(rows):
        for col in range(cols):
            if grid[row][col] == 1:
                queue = deque([(row, col)])
                level = 0

                while queue:
                    for _ in range(len(queue)):
                        r, c = queue.popleft()
                        for dr, dc in directions:
                            nr, nc = r + dr, c + dc
                            if nr < 0 or nr >= rows or nc < 0 or nc >= cols or grid[nr][nc] != walk:
                                continue
                            grid[nr][nc] -= 1               # mark visited
                            distance[nr][nc] += level + 1   # accumulate distance
                            answer = min(answer, distance[nr][nc])
                            queue.append((nr, nc))
                    level += 1

                walk -= 1

    return -1 if answer == float('inf') else answer


if __name__ == "__main__":
    import copy

    test_cases = [
        ([[1,0,2,0,1],[0,0,0,0,0],[0,0,1,0,0]], 7),
        ([[1,0]],                                 1),
        ([[1]],                                  -1),
        ([[0,1,0],[0,0,0],[0,1,0]],               2),
        ([[0,2,1]],                              -1),
    ]

    for grid, expected in test_cases:
        result = shortest_distance(copy.deepcopy(grid))
        status = "PASS" if result == expected else "FAIL"
        print(f"[{status}] result={result} expected={expected}")
